doc/developers.md

```diff
@@ -86,3 +97,46 @@ spdlog::debug("Send help");
 
 The C++ `main` function is in openandorder.cpp <!-- words with a dot in them are assumed to be files so this will hyperlink to openandorder.cpp iff *that* file is also documented. --> however this only really does file I/O and setup.
 The main FDTD algorithm code is in iterator.cpp <!-- won't be linked as an undocumented file doesn't exist for doxygen... this is fine, we can link to the real file in github.--> and classes included therein.
+
+## Testing
+
+We have two [levels of tests](https://en.wikipedia.org/wiki/Software_testing#Testing_levels): unit tests, and full system tests.
+
+### Unit 
+The unit tests use [catch2](https://github.com/catchorg/Catch2/blob/devel/docs/Readme.md#top) macros. See [tests/unit](https://github.com/UCL/TDMS/blob/main/tdms/tests/unit) for good examples in the actual test code, but as a rough sketch you need:
+
+```{.cpp}
+#include <catch2/catch_test_macros.hpp>
+#include "things_to_be_tested.h"
+
+TEST_CASE("Write a meaningful test case name") {
+    // set up function calls or whatever
+    REQUIRE_THROW(<something>)
+    CHECK(<something>)
+}
+```
+To run the unit tests, [compile](#compiling) with `-DBUILD_TESTING=ON` and then run `ctest` from the build directory or execute the test executable `./tdms_tests`.
+
+It's good practice, and reassuring for your pull-request reviewers, if new C++ functionality is at covered by unit tests.
+
+### System {#system-tests}
+
+The full system tests are written in Python 3, and call the `tdms` executable for known inputs and compare to expected outputs.
+We use [pytest](https://docs.pytest.org) and our example data is provided as zip files on [zenodo](https://zenodo.org/). 
+
+There are a few [python packages you will need](https://github.com/UCL/TDMS/blob/main/tdms/tests/requirements.txt) before running the tests so run:
+```{.sh}
+pip3 install -r tdms/tests/requirements.txt
+```
+if you don't already have them.
+
+When you run the tests for the first time, the example data will be downloaded to `tdms/tests/system/data` (which is [ignored by git](https://github.com/UCL/TDMS/blob/main/.gitignore)).
+Subsequent runs of the test will not redownload unless you manually delete the zip file.
+
+A good example of running the `tdms` executable for a given input and expected output is [test_arc01.py](https://github.com/UCL/TDMS/blob/main/tdms/tests/system/test_arc01.py)
+
+You need to [compile](#compiling) `tdms`, then the system tests can be run, e.g. from the build directory:
+
+```{.sh}
+py.test ../tests/system/
```

Comment:
```suggestion
pytest ../tests/system/
```
I think invoking `py.test` (as opposed to `pytest`) has been discouraged for a few years now.

Reply:
Brainless copy-over. And a good spot.